sklearn/preprocessing/imputation.py

```diff
@@ -59,6 +59,95 @@ def _most_frequent(array, extra_value, n_repeat):
+        mask_X = _get_mask(X, value_to_mask)
+        XT = np.transpose(X)
+        N = (~mask_X) * 1
+        NT = np.transpose(N)
+        X[mask_X] = 0
+
+        # Matrix formula to calculate pair-wise distance between all vectors in a
+        # matrix with missing values. It zero-weights coordinates with missing value
+        # in either vector in the pair and up-weights the remaining coordinates.
+        # Matrix formula derived by: Shreya Bhattarai <shreya.bhattarai@gmail.com>
+
+        """
+        Store np.dot(N, (XT * XT)) and add its transpose rather than 
+        redoing a matrix product
+        dist = np.sqrt((X.shape[1] * 1 / ((np.dot(N, NT)))) * (
+            np.dot(N, (XT * XT)) - 2 * (np.dot(X, XT)) +
+            np.dot((X * X), NT)))        
+
+        N_dot_XT2 = np.dot(N, (XT * XT))
+        N_dot_XT2_T = np.transpose(N_dot_XT2)
+        """
+
+        N_dot_XT2 = np.dot(N, (XT * XT))
+        N_dot_XT2_T = np.transpose(N_dot_XT2)
+
+        dist = np.sqrt((X.shape[1] * 1 / ((np.dot(N, NT)))) * (
+            N_dot_XT2 - 2 * (np.dot(X, XT)) +
+            N_dot_XT2_T))
+
+        # Set distance with self to np.inf
+        np.fill_diagonal(dist, np.inf)
+
+    else:
+        # ValueError if X and Y have incompatible dimensions
+        if X.shape[1] != Y.shape[1]:
+            raise ValueError("The search dimension of the matrices "
+                             "are not equal: [{0}] versus [{1}]".
+                             format(X.shape[1], Y.shape[1]))
+
+        mask_Y = _get_mask(Y, value_to_mask)
+        NY = (~mask_Y) * 1
+        YT = np.transpose(Y)
+        mask_YT = _get_mask(YT, value_to_mask)
+        NYT = np.transpose(NY)
+        YT[mask_YT] = 0
+
+        NX = (~mask_X) * 1
+        X[mask_X] = 0
+
+        # Matrix formula to calculate pair-wise distance between all vectors in a
+        # matrix X to vectors in matrix Y. It handles missing values the same way
+        # as for a single matrix.
+        # Matrix formula derived by: Shreya Bhattarai <shreya.bhattarai@gmail.com>
+
+        dist = np.sqrt((X.shape[1] * 1 / ((np.dot(NX, NYT)))) *
+                       (np.dot((X * X), NYT) - 2 * (np.dot(X, YT)) + np.dot(NX, (YT * YT))))
+
+    # Ensure enough candidate neighbors are available
+    n_candidates = dist.shape[1] if Y is not None else dist.shape[1] - 1
+    if n_candidates < n_neighbors:
+        raise ValueError("There are only %d candidate neighbors, "
+                         "but n_neighbors=%d."
+                         % (dist.shape[1] - 1, n_neighbors))
+
+    # Missing locations and counts
+    row_missing_sum_X = mask_X.sum(axis=1)
+    # is_row_missing_X = np.any(mask_X, axis=1)
+    # is_col_missing_X = np.any(mask_X, axis=0)
+    col_missing_index_X = np.where(mask_X)[1]
+
+    # Arg-partition (quasi-argsort) of n_neighbors and retrieve them
+    nbors_index = np.argpartition(dist, n_neighbors - 1, axis=1)
+    knn_row_index = nbors_index[:, :n_neighbors]
```

Comment:
As far as I can remember, the package referenced above does the following: 
i) if some of the neighbors have that feature not missing then only those neighbors are used for that particular feature
ii) if non of the neighbors have non-missing for a particular feature then it uses the column/feature mean
I will have a look at their source code when I start working on this.